File: example/det_by_minor.cpp
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later
// SPDX-FileCopyrightText: Bradley M. Bell <bradbell@seanet.com>
// SPDX-FileContributor: 2023 Bradley M. Bell
// ----------------------------------------------------------------------------

/*
{xrst_begin det_by_minor.cpp}

det_by_minor: Example and Test
##############################

{xrst_literal
   // BEGIN C++
   // END C++
}

{xrst_end det_by_minor.cpp}
*/
// BEGIN C++
# include <boost/test/unit_test.hpp>
# include <cmpad/det_by_minor.hpp>

# define BOOST_TEST_DYN_LINK
# define BOOST_TEST_MODULE example

BOOST_AUTO_TEST_CASE(det_by_minor)
{
   // size
   // dimension of the matrix
   size_t ell = 3;
   //
   // det
   cmpad::det_by_minor<double> det;
   //
   // det.setup
   cmpad::option_t option;
   option.size = ell;
   det.setup(option);
   //
   // x
   cmpad::vector<double>  x = {
      1., 2., 3.,  // x[0] x[1] x[2]
      3., 2., 1.,  // x[3] x[4] x[5]
      2., 1., 2.   // x[6] x[7] x[8]
   };
   BOOST_REQUIRE( x.size() == ell * ell );
   //
   // y
   const cmpad::vector<double>& y = det(x);
   BOOST_CHECK( y.size() == 1 );
   //
   // check
   double check;
   check = x[0]*(x[4]*x[8] - x[5]*x[7])
         - x[1]*(x[3]*x[8] - x[5]*x[6])
         + x[2]*(x[3]*x[7] - x[4]*x[6]);
   BOOST_CHECK(y[0] == check);
}

// END C++

File: python/cmpad/det_by_minor.py
# SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later
# SPDX-FileCopyrightText: Bradley M. Bell <bradbell@seanet.com>
# SPDX-FileContributor: 2023 Bradley M. Bell
# ---------------------------------------------------------------------------
r'''
{xrst_begin_parent py_det_by_minor}

Python Determinant Using Expansion by Minors
############################################

Syntax
******
| |tab| *det* = ``cmpad.det_by_minor()``
| |tab| *det* . ``setup`` ( *option* )
| |tab| *y* = *det* ( *x* )

Prototype
*********
{xrst_literal
   # BEGIN PROTOTYPE
   # END PROTOTYPE
}

Algorithm
*********
This is a Python implementation of the
det_by_minor :ref:`det_by_minor@Algorithm` .

Scalar
******
We use *Scalar* for the type of the elements of *x* and *y*.

det
***
The :ref:`py_fun_obj-name` *det* computes the determinant of a square matrix.

ell
***
see :ref:`det_by_minor@ell` .

n_arg
*****
see :ref:`det_by_minor@option@n_arg` .

n_other
=======
see :ref:`det_by_minor@option@n_other` .

x
*
The argument *x* has size *n_arg* = :math:`\ell * \ell` .
The elements of the matrix :math:`A(x)` is defined as follows:
for :math:`i = 0 , \ldots , \ell-1` and :math:`j = 0 , \ldots , \ell-1`, by

.. math::

   A(x)_{i,j} = x[ i * \ell + j]

y
*
The return value *y* has length one and its element
is equal to the determinant of :math:`A(x)`.

{xrst_toc_hidden after
   python/xam/det_by_minor.py
   python/cmpad/det_of_minor.py
}

Example
*******
The file
:ref:`xam_det_by_minor.py-name`
contains an example and test of ``det_by_minor`` .

Source Code
***********
:ref:`det_by_minor.py-name` displays the source code for this algorithm.

{xrst_end py_det_by_minor}
---------------------------------------------------------------------------
{xrst_begin det_by_minor.py}

Python det_by_minor: Source Code
################################
{xrst_literal
   # BEGIN PYTHON
   # END PYTHON
}

{xrst_end det_by_minor.py}
---------------------------------------------------------------------------
'''
# BEGIN PYTHON
import math
import cmpad
# BEGIN PROTOTYPE
class det_by_minor :
   #
   def option(self) :
      return self.option
   #
   def domain(self) :
      return self.option['n_arg']
   #
   def range(self) :
      return 1
   #
   def setup(self, option) :
      assert type(option) == dict
      assert type( option['n_arg'] ) == int
      assert type( option['n_other'] ) == int
      assert option['n_arg'] > 0
      assert option['n_other'] == 0
      # END PROTOTYPE
      #
      # option
      self.option = option
      #
      # self.ell
      ell = int( math.sqrt( option['n_arg'] ) )
      if( ell * ell < option['n_arg'] ) :
         ell += 1
      assert ell * ell == option['n_arg']
      self.ell = ell
      #
      # r, c
      self.r = (ell + 1) * [0]
      self.c = (ell + 1) * [0]
      for i in range(ell) :
         self.r[i] = i + 1
         self.c[i] = i + 1
      self.r[ell] = 0
      self.c[ell] = 0
   #
   # call
   def __call__(self, x) :
      #
      # d
      d = cmpad.det_of_minor(x, self.ell, self.ell, self.r, self.c)
      #
      return [ d ]

# END PYTHON
